device/common/include/traccc/seeding/device/impl/count_triplets.ipp

```diff
@@ -0,0 +1,163 @@
+
+    // Apply the conformal transformation to middle-bot doublet
+    traccc::lin_circle lb = doublet_finding_helper::transform_coordinates(spM, spB, true);
+
+    // Calculate some physical quantities required for triplet compatibility check 
+    scalar iSinTheta2 = 1 + lb.cotTheta() * lb.cotTheta();
+    scalar scatteringInRegion2 = m_config.maxScatteringAngle2 * iSinTheta2;
+    scatteringInRegion2 *=
+        m_config.sigmaScattering * m_config.sigmaScattering;
+    scalar curvature, impact_parameter;
+
+    // find the reference (start) index of the mid-top doublet container
+    // item vector, where the doublets are recorded The start index is
+    // calculated by accumulating the number of mid-top doublets of all
+    // previous compatible middle spacepoints
+    unsigned int mb_end_idx = 0;
+    unsigned int mt_start_idx = 0;
+    unsigned int mt_end_idx = 0;
+    for (unsigned int i = 0; i < num_compat_spM_per_bin; ++i) {
+        mb_end_idx += doublet_counter_per_bin[i].m_nMidBot;
+        mt_end_idx += doublet_counter_per_bin[i].m_nMidTop;
+
+        if (mb_end_idx > mb_idx) {
+            break;
+        }
+        mt_start_idx += doublet_counter_per_bin[i].m_nMidTop;
+    }
+    
+    if (mt_end_idx >= mid_top_doublets_per_bin.size()) {
+            mt_end_idx =
+                fmin(mid_top_doublets_per_bin.size(), mt_end_idx);
+        }
+
+        if (mt_start_idx >= mid_top_doublets_per_bin.size()) {
+            return;
+        }
+    
+    // number of triplets per thread (or per middle-bot doublet)
+    unsigned int num_triplets_per_mb = 0;
+
+    // iterate over mid-top doublets
+    for (unsigned int i = mt_start_idx; i < mt_end_idx; ++i) {
+        const traccc::doublet& mid_top_doublet = mid_top_doublets_per_bin[i];
+
+        const unsigned int& spT_idx = mid_top_doublet.sp2.sp_idx;
+        const traccc::internal_spacepoint<traccc::spacepoint>& spT = internal_sp_device.bin(bin_idx)[spT_idx];
+
+        // Apply the conformal transformation to middle-top doublet
+        traccc::lin_circle lt =
+            doublet_finding_helper::transform_coordinates(spM, spT, false);
+
+        // Check if mid-bot and mid-top doublets can form a triplet
+        if (triplet_finding_helper::isCompatible(
+                spM, lb, lt, m_config, iSinTheta2, scatteringInRegion2,
+                curvature, impact_parameter)) {
+            num_triplets_per_mb++;
+        }
+    }
+
+    // if the number of triplets per mb is larger than 0, write the triplet
+    // counter into the container
+    if (num_triplets_per_mb > 0) {
+        vecmem::device_atomic_ref<uint32_t> obj(num_compat_mb_per_bin);
+        const int pos = obj.fetch_add(1);
+        vecmem::device_atomic_ref<uint32_t> objT(num_triplets);
+        objT.fetch_add(num_triplets_per_mb);
+        triplet_counter_at_bin[pos] = {mid_bot.sp1.sp_idx,
+                                        num_triplets_per_mb};
```

Comment:
:thinking: It should be possible to do this a little more elegantly...

First off, you should not use `uint32_t` to refer to `unsigned int` variables. They are the same in practice, but let's use consistent type names throughout the code.

But the more interesting thing is how the triplet counter container is being filled. :thinking: We should try to do that with a `push_back(...)` function. You see, right now the `num_compat_mb_per_bin` variable is just there to help choose the correct bin in the `triplet_counter_at_bin` vector. But we could make that jagged vector "resizable", and just use the `vecmem::device_vector<T>::push_back(...)` function on it to add a new element into it.

But since this will require a bit more tweaks in the code, for now let's just replace `uint32_t` with `unsigned int`. That should be enough for this PR.